src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala

```diff
@@ -7,159 +7,150 @@ import akka.http.scaladsl.server.Directives._
-    provide(CreateWorkbenchUser(genWorkbenchUserId(System.currentTimeMillis()), oidcHeaders.googleSubjectId, oidcHeaders.email, None))
+  private def externalIdFromHeaders: Directive1[Either[GoogleSubjectId, AzureB2CId]] = headerValueByName(userIdHeader).map { idString =>
+    Try(BigInt(idString)).fold(
+      _ => Right(AzureB2CId(idString)),    // could not parse id as a Long, treat id as b2c id which are uuids
+      _ => Left(GoogleSubjectId(idString)) // id is a number which is what google subject ids look like
+    )
+  }
 }
 
 object StandardUserInfoDirectives {
-  val SAdomain: Regex = "\\S+@\\S+\\.iam\\.gserviceaccount\\.com".r
+  val SAdomain: Regex = "(\\S+@\\S+\\.iam\\.gserviceaccount\\.com$)".r
   val accessTokenHeader = "OIDC_access_token"
   val expiresInHeader = "OIDC_CLAIM_expires_in"
   val emailHeader = "OIDC_CLAIM_email"
-  val googleSubjectIdHeader = "OIDC_CLAIM_user_id"
-  val authorizationHeader = "Authorization"
-
-  implicit val jwtPayloadFormat: JsonFormat[JwtUserInfo] = jsonFormat2(JwtUserInfo)
-
-  def isServiceAccount(email: WorkbenchEmail): Boolean =
-    SAdomain.pattern.matcher(email.value).matches
-
-  def getUserInfoFromOidcHeaders(directoryDAO: DirectoryDAO, oidcHeaders: OIDCHeaders, samRequestContext: SamRequestContext): IO[UserInfo] =
-    Try(oidcHeaders.expiresIn.toLong).fold(
-      t => IO.raiseError(new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"expiresIn ${oidcHeaders.expiresIn} can't be converted to Long because  of $t"))),
-      expiresInLong => getUserInfo(oidcHeaders.token, oidcHeaders.googleSubjectId, oidcHeaders.email, expiresInLong, directoryDAO, samRequestContext)
-    )
-
-  def getUserInfo(token: OAuth2BearerToken, googleSubjectId: GoogleSubjectId, email: WorkbenchEmail, expiresIn: Long, directoryDAO: DirectoryDAO, samRequestContext: SamRequestContext): IO[UserInfo] =
-    if (isServiceAccount(email)) {
-      // If it's a PET account, we treat it as its owner
-      directoryDAO.getUserFromPetServiceAccount(ServiceAccountSubjectId(googleSubjectId.value), samRequestContext).flatMap {
-        case Some(pet) => IO.pure(UserInfo(token, pet.id, pet.email, expiresIn.toLong))
-        case None => lookUpByGoogleSubjectId(googleSubjectId, directoryDAO, samRequestContext).map(uid => UserInfo(token, uid, email, expiresIn))
-      }
-    } else {
-      lookUpByGoogleSubjectId(googleSubjectId, directoryDAO, samRequestContext).map(uid => UserInfo(token, uid, email, expiresIn))
+  val userIdHeader = "OIDC_CLAIM_user_id"
+  val idpAccessTokenHeader = "OAUTH2_CLAIM_idp_access_token"
+
+  def getUserInfo(directoryDAO: DirectoryDAO, maybeGoogleOpaqueTokenResolver: Option[GoogleOpaqueTokenResolver], oidcHeaders: OIDCHeaders, samRequestContext: SamRequestContext): IO[UserInfo] = {
+    oidcHeaders match {
+      case OIDCHeaders(token, Left(googleSubjectId), _, saEmail@WorkbenchEmail(SAdomain(_)), _) =>
```

Comment:
For these pattern matches, I'm wondering if it wouldn't be better to write it as:
```suggestion
      case OIDCHeaders(_, Left(googleSubjectId), _, saEmail@WorkbenchEmail(SAdomain(_)), _) =>
```
And same thing in the other two `case`s to use `_` for the headers we don't care about in the pattern match?  I believe it would make it easier to read if the only things the patterns are checking are the fields we care about, which seems to be the `externalId` and `email`.

Reply:
the change is easy enough to make here and I will do so to make the cases clear. However in general this is not always desired because this is not only pattern matching but extracting values out of the pattern. So in this case I will have a bunch more `oidcHeaders.` but in general it can be a pain.